sql: make `SHOW TRACE FOR <QUERY>` work with distsql

asubiotto opened the pull request:
Closes #16562

Previously, to trace distSQL queries, a user would have had to set
session tracing on and then run `SHOW TRACE FOR SESSION`, ignoring
unrelated output. The reason for this limitation is that the
`showTraceNode` wraps the plan with a bunch of distSQL-incompatible
nodes and so the whole query is run through local execution.

The second commit leverages the distSQLNode introduced in the first commit to run the distSQL-compatible subplan through distSQL.

I still have a couple of `TODO`s with some questions and would appreciate a double-check on all the stuff I added with the `distSQLNode` since I am not very familiar with the planning code.

Commit:
sql: add distSQLWrapper

This planNode uses the DistSQLPlanner to plan and run a distsql query.
This allows us to run a sub-plan of a local plan through distSQL.
The motivation for this work is to be able to run `SHOW TRACE FOR
<QUERY>` and execute `<QUERY>` with distSQL. This cannot be done
currently because there are some distSQL-incompatible nodes in the
resulting plan.

This node might also prove useful for other cases.

Release note: None

Commit:
sql: make `SHOW TRACE FOR <QUERY>` work with distsql

Closes #16562

Previously, to trace distSQL queries, a user would have had to set
session tracing on and then run `SHOW TRACE FOR SESSION`, ignoring
unrelated output. The reason for this limitation is that the
`showTraceNode` wraps the plan with a bunch of distSQL-incompatible
nodes and so the whole query is run through local execution.

This commit leverages the distSQLWrapper to run the distSQL-compatible
subplan through distSQL.

Release note (sql change): `SHOW TRACE FOR <QUERY>` now runs <QUERY>
through distSQL if supported. `SHOW KV TRACE FOR <QUERY>` still runs
<QUERY> through local SQL.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24709)
<!-- Reviewable:end -->

asubiotto commented:
There are some failing tests I still have to deal with but wanted to get this out and discussed.

knz commented:
Thanks alfonso for this preliminary patch.

Overall I see while this would work, but I am not too fond of the approach. It seems to me that the distsql physical planner should take ownership of the problem instead, and schedule showTrace as a dedicated processor.

Supposing we keep your approach I'd like to see more clarity on the following points:

1) the mechanism proposed here, while sufficient to solve the proximate show trace problem, is fundamentally unable to return rows to a client. I think that the name "distSQLNode" is promising more than it can deliver. May I suggest "distSQLTraceWrapperNode" or something similar.

2) please check what happens with `show trace for select * from [show trace for table kv]`

3) please check what happens with `select * from [show trace for table kv], [show trace for table abc]`

---

Reviewed 9 of 9 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/distsql_node.go, line 47 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9ssSApC-AJAEzvsDoU:-L9ssSApC-AJAEzvsDoV:b9i15bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L47)):*
> ```Go
> 
> func (n *distSQLNode) Next(params runParams) (bool, error) {
> 	if !n.executed {
> ```

`if n.executed { return false, nil }`

then the rest on the main path.

---

*[pkg/sql/distsql_node.go, line 81 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9ssdpKEX7OEaoxher4:-L9ssdpKEX7OEaoxher5:brc1kuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L81)):*
> ```Go
> 
> func (n *distSQLNode) Close(ctx context.Context) {
> 	if n.plan != nil {
> ```

Why do you need to check n.plan != nil? I don't see it being set to nil anywhere.

---

*[pkg/sql/expand_plan.go, line 127 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9sr0ktDf8wgKOYU8s3:-L9sr0ktDf8wgKOYU8s4:b-2g9e7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/expand_plan.go#L127)):*
> ```Go
> 	case *distSQLNode:
> 		distSQLParams := noParamsBase
> 		distSQLParams.atTop = true
> ```

I do not understand why you are setting atTop here. My intuition (and the spec for the field) suggest this is incorrect. See my other comment in show_trace.go. I think the call to doExpandPlan here should be done as usual on the child node, and then after it returns, check whether the query can be run with distsql and insert the new node type to deal with execution.

---

*[pkg/sql/expand_plan.go, line 129 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9sskV25aw-l2sPBWpo:-L9sskV25aw-l2sPBWpp:b-kh7u06) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/expand_plan.go#L129)):*
> ```Go
> 		distSQLParams.atTop = true
> 		n.plan, err = doExpandPlan(ctx, p, distSQLParams, n.plan)
> 		if err != nil {
> ```

This test on `err` is not needed. IT is only needed in each of the `case` of this switch if there is some code afterwards in the same `case`. See `deleteNode` above, etc.

---

*[pkg/sql/expand_plan.go, line 149 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9st8NU7tCRAvg8dQzi:-L9st8NU7tCRAvg8dQzj:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/expand_plan.go#L149)):*
> ```Go
> 		showTraceParams.atTop = true
> 		n.plan, err = doExpandPlan(ctx, p, showTraceParams, n.plan)
> 		if err != nil {
> ```

ditto

---

*[pkg/sql/expand_plan.go, line 155 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9st9EhAYhQDcLrHPZQ:-L9st9EhAYhQDcLrHPZR:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/expand_plan.go#L155)):*
> ```Go
> 	case *showTraceReplicaNode:
> 		n.plan, err = doExpandPlan(ctx, p, noParams, n.plan)
> 		if err != nil {
> ```

ditto

---

*[pkg/sql/show_trace.go, line 118 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9stt_S9iqUP2WKwxyp:-L9stt_S9iqUP2WKwxyq:b-4rf5gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/61334d2e7c5147595da03f7a083bc2f8cf7f9c41/pkg/sql/show_trace.go#L118)):*
> ```Go
> 		return nil, err
> 	}*/
> 	if err := p.makePlan(ctx, Statement{AST: n.Statement}); err != nil {
> ```

oy, no. makePlan can't be called here. The interface and the contracts simply don't allow this.
Just keep newPlan() as was done before and see my comment below for the remainder of the solution.

---

*[pkg/sql/show_trace.go, line 141 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9suEFM6xvBElCXEaZa:-L9suEFM6xvBElCXEaZb:b-dvsl9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/61334d2e7c5147595da03f7a083bc2f8cf7f9c41/pkg/sql/show_trace.go#L141)):*
> ```Go
> 	// populated on a scanNode and the recommendation for spans != 1 is
> 	// canDistribute which results in useDistSQL = false.
> 	if useDistSQL, err := p.ExecCfg().DistSQLPlanner.CheckSupport(stmtPlan); useDistSQL && err == nil {
> ```

This code should not be here. It is too early to perform these checks. Instead, it should be done during expandPlan, on the return path from the recursion on the child query. See my suggestion in expand_plan.go.

---

*[pkg/sql/walk.go, line 591 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9sqm18C3qo6jLR_D-0:-L9sqm18C3qo6jLR_D-1:bnvzk0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/walk.go#L591)):*
> ```Go
> 	reflect.TypeOf(&dropSequenceNode{}):         "drop sequence",
> 	reflect.TypeOf(&DropUserNode{}):             "drop user | role",
> 	reflect.TypeOf(&distSQLNode{}):              "dist_sql query",
> ```

I don't think the underscore here (and below) is warranted. You can also replace ` | ` by `/` above.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-:-L9srFiv7MbwFNNwTJKi:bo9ypp4)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I suggested this approach. The problem with making a `showTrace` processor is that it'd have to replicate the window function and whatever we use in the query rewrite that `ShowTrace()` performs, and I'm not sure how to do that.
I'd rather keep this wrapper but, as you say, make it able to return rows.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_node.go, line 27 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9u_khl38D7GhUnx8I2:-L9u_khl38D7GhUnx8I3:b95geeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L27)):*
> ```Go
> // planning and execution. The plan is assumed to be supported by distSQL.
> // Rows are discarded and not returned.
> type distSQLNode struct {
> ```

I think `distSQLWrapper` would carry more meaning

---

*[pkg/sql/distsql_node.go, line 39 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9u_veM9xfbwKwAhl2l:-L9u_veM9xfbwKwAhl2m:bb9fnpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L39)):*
> ```Go
> // plan: The wrapped execution plan to be physically planned and executed.
> // stmtType: The statement type of the statement represented by plan.
> func (p *planner) makeDistSQLNode(plan planNode, stmtType tree.StatementType) planNode {
> ```

`newDistSQLNode` is more accurate. Unless this is the pattern everywhere.

---

*[pkg/sql/distsql_node.go, line 48 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9uaKlw4m5UqbBT3p00:-L9uaKlw4m5UqbBT3p01:bdr4cv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L48)):*
> ```Go
> func (n *distSQLNode) Next(params runParams) (bool, error) {
> 	if !n.executed {
> 		// For now, the distSQLNode discards all result rows since it is only
> ```

well as long as you're doing this work, consider doing it for realz - accumulate results in a buffer, push back on the producer when the buffer is full, synchronize with the buffer on Next() calls

---

*[pkg/sql/show_trace.go, line 113 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9ubuh72zM1SyGJ2CjH:-L9ubuh72zM1SyGJ2CjI:btfvlc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/61334d2e7c5147595da03f7a083bc2f8cf7f9c41/pkg/sql/show_trace.go#L113)):*
> ```Go
> 
> 	// SHOW TRACE FOR SELECT ...
> 	/*stmtPlan, err := p.newPlan(ctx, n.Statement, nil)
> ```

dead code

---

*[pkg/sql/show_trace.go, line 141 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9suEFM6xvBElCXEaZa:-L9ucXtYCI9-zksXkVxU:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/61334d2e7c5147595da03f7a083bc2f8cf7f9c41/pkg/sql/show_trace.go#L141)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This code should not be here. It is too early to perform these checks. Instead, it should be done during expandPlan, on the return path from the recursion on the child query. See my suggestion in expand_plan.go.
</blockquote></details>

+1

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-:-L9ucvN8BSRywprQ7Xiu:b9fkhr8)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_node.go, line 27 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9u_khl38D7GhUnx8I2:-L9uh0KTBNdA7BphJVTb:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L27)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think `distSQLWrapper` would carry more meaning
</blockquote></details>

:+1:

---

*[pkg/sql/distsql_node.go, line 48 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709#-L9uaKlw4m5UqbBT3p00:-L9uh3zLD109l8RE9Szg:b-fiq7it) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well as long as you're doing this work, consider doing it for realz - accumulate results in a buffer, push back on the producer when the buffer is full, synchronize with the buffer on Next() calls
</blockquote></details>

be careful with memory usage though

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsql_node.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-L9uaKlw4m5UqbBT3p00:-LBMD3eOAtq1qNlXLk-w:b-vwoboq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L48)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ok thanks!
</blockquote></details>

you push back on a producer by blocking in the `Push()` call to its consumer - in this case the consumer is the `distSQLReceiver` and you can block in the results writer.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsql_node.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-L9uaKlw4m5UqbBT3p00:-LBR5hK81iHizqi3G6vp:b-xndb03) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

you push back on a producer by blocking in the `Push()` call to its consumer - in this case the consumer is the `distSQLReceiver` and you can block in the results writer.
</blockquote></details>

But is that sufficient? It may be for general cases but I'm wondering about scenarios in which blocking the `Push()` call doesn't actually block processing. Take the existence of a `hashRouter` in a plan for example. I think it would keep on buffering rows.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsql_node.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-L9uaKlw4m5UqbBT3p00:-LBRQGgGD1jOi0EWuhhB:b-3yu9a9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e0410daf0514fd665768edf8769f5cf8ebcb18/pkg/sql/distsql_node.go#L48)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

But is that sufficient? It may be for general cases but I'm wondering about scenarios in which blocking the `Push()` call doesn't actually block processing. Take the existence of a `hashRouter` in a plan for example. I think it would keep on buffering rows.
</blockquote></details>

The `hashRouter` blocks when all its outputs are blocked. Not consuming input is generally the pushback mechanism in DistSQL; if that doesn't work, we have a general problem. This consumer here is no different from any other. The regular `distSQLReceiver` used for running distsql queries uses a results writer that flushes to the network socket - which is blocking.
There's a `RowResultWriter` somewhere that wraps a `rowContainer`. You might be able to use that.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 8 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/executor.go, line 2092 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbF5YVDaBQ2KZHuveI:-LBbF5YVDaBQ2KZHuveJ:baqfcvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e632c961b23cbf5f48f364b46d6510ba9f7341/pkg/sql/executor.go#L2092)):*
> ```Go
> // given logical plan, based on the session settings. If plan is nil,
> // planner.curPlan.plan is used as plan.
> func shouldUseDistSQL(
> ```

it's really confusing that this takes both a planner and a plan. If possible, I think it'd be a lot better if it only took `plan`. I see that `planner` is used for the `planner.setUnlimited(plan)` call, but I don't know what that's about; I hope it doesn't belong in this function and that callers can do it when/if needed.
cc @knz @raduberinde who may know about that call - what does it have to do with `shouldUseDistSQL()`?

---

*[pkg/sql/opt_limits.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbEVeB3SLJnG_Vlpjf:-LBbEVeC38RmDdJ74IW9:b-g2mar3) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/opt_limits.go#L174)):*
> ```Go
> 		}
> 	case *distSQLWrapper:
> 		p.setUnlimited(n.plan)
> ```

is this the right thing to do here? Shouldn't we propagate the limit to `n.plan`?

---

*[pkg/sql/distsql_wrapper.go, line 25 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbCxrs8k6k4iwoS93g:-LBbCxrs8k6k4iwoS93h:b1e8ku) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/distsql_wrapper.go#L25)):*
> ```Go
> )
> 
> const distSQLWrapperResultChanSize = 1024
> ```

nit: `distSQLWrapperBufferSize`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Thanks for the changes!

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-:-LBbFnEdDyzWnL2ZZBbV:baeznli)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 8 files at r5, 1 of 5 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/executor.go, line 2092 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbF5YVDaBQ2KZHuveI:-LBbH3sADoJflFclUjXt:bebph1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e632c961b23cbf5f48f364b46d6510ba9f7341/pkg/sql/executor.go#L2092)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it's really confusing that this takes both a planner and a plan. If possible, I think it'd be a lot better if it only took `plan`. I see that `planner` is used for the `planner.setUnlimited(plan)` call, but I don't know what that's about; I hope it doesn't belong in this function and that callers can do it when/if needed.
cc @knz @raduberinde who may know about that call - what does it have to do with `shouldUseDistSQL()`?
</blockquote></details>

yes good question.

---

*[pkg/sql/opt_limits.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbEVeB3SLJnG_Vlpjf:-LBbH0Zz40BTpSAUQGu-:b-9qyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/opt_limits.go#L174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is this the right thing to do here? Shouldn't we propagate the limit to `n.plan`?
</blockquote></details>

I believe this is correct.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt_limits.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbEVeB3SLJnG_Vlpjf:-LBbHbaU9nDuuMwubIGr:blxu20w) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/opt_limits.go#L174)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I believe this is correct.
</blockquote></details>

why not propagate?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt_limits.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbEVeB3SLJnG_Vlpjf:-LBbXzwDBFVuHSXWTtj9:b-5x78bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/opt_limits.go#L174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why not propagate?
</blockquote></details>

Think about the following scenario:

1. user types `EXPLAIN SELECT * FROM kv WHERE v = 3`. output contains two rows: a `filter` stage and a `scan` stage below this.
2. user thinks "oh great now I can make a query that extracts just the parameters of the scan". So they type, correctly, `SELECT * FROM [EXPLAIN SELECT * FROM kv WHERE v = 3] OFFSET 2 LIMIT 1`
3. now watch what happens:

   - if we propagate the outer limit inside, now the _inner_ query becomes `SELECT * FROM kv WHERE v = 3 OFFSET 2 LIMIT 1` and the `scan` parameters change (they appear to have a limit, whereas the original query did not specify any limit).

   - if we don't propagate the outer limit inside, the inner query is unchanged. We're good.

So no, don't propagate the limit.

FWIW, the other code for explain (explainNode, explainDsitSQLNode) doesn't do it either.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Fixed nits and copying rows in `AddRow` now because they are reused by the `distsSQLReceiver`.

---

Review status: 3 of 14 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/distsql_wrapper.go, line 25 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbCxrs8k6k4iwoS93g:-LBc6xgc7-9kh02xdnhj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/distsql_wrapper.go#L25)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: `distSQLWrapperBufferSize`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-:-LBcETti2_ajIQrD-z_G:b33fp6y)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 14 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt_limits.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbEVeB3SLJnG_Vlpjf:-LBcF5NtBjEPzcVawfd1:bw50d1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/opt_limits.go#L174)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Think about the following scenario:

1. user types `EXPLAIN SELECT * FROM kv WHERE v = 3`. output contains two rows: a `filter` stage and a `scan` stage below this.
2. user thinks "oh great now I can make a query that extracts just the parameters of the scan". So they type, correctly, `SELECT * FROM [EXPLAIN SELECT * FROM kv WHERE v = 3] OFFSET 2 LIMIT 1`
3. now watch what happens:

   - if we propagate the outer limit inside, now the _inner_ query becomes `SELECT * FROM kv WHERE v = 3 OFFSET 2 LIMIT 1` and the `scan` parameters change (they appear to have a limit, whereas the original query did not specify any limit).

   - if we don't propagate the outer limit inside, the inner query is unchanged. We're good.

So no, don't propagate the limit.

FWIW, the other code for explain (explainNode, explainDsitSQLNode) doesn't do it either.
</blockquote></details>

I may well be confused - but here we're not talking about explain or show trace propagating limits. We're talking about this generic `distSQLWrapper` propagating them. If the wrapper has a limit, the inner plan has the same limit. In the case of `SHOW TRACE`, the limit should not be propagated to the wrapper.
No?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 3 of 14 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 2092 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbF5YVDaBQ2KZHuveI:-LCUCH9N1xOcfEjFGBli:b-ck06f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e632c961b23cbf5f48f364b46d6510ba9f7341/pkg/sql/executor.go#L2092)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

yes good question.
</blockquote></details>

ping @RaduBerinde

---

*[pkg/sql/opt_limits.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbEVeB3SLJnG_Vlpjf:-LCUIoThBMdSBxZ8UxVC:b-8b822u) ([raw file](https://github.com/cockroachdb/cockroach/blob/857bd3707587dd7e026fa00cb05ff15abaa47e72/pkg/sql/opt_limits.go#L174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I may well be confused - but here we're not talking about explain or show trace propagating limits. We're talking about this generic `distSQLWrapper` propagating them. If the wrapper has a limit, the inner plan has the same limit. In the case of `SHOW TRACE`, the limit should not be propagated to the wrapper.
No?
</blockquote></details>

@andreimatei, I believe you're right. Changed this.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/executor.go, line 2092 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbF5YVDaBQ2KZHuveI:-LCUursd1M5Qqa_Nis3p:bqg4vs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e632c961b23cbf5f48f364b46d6510ba9f7341/pkg/sql/executor.go#L2092)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

ping @RaduBerinde
</blockquote></details>

The distsql planner did not run plans with limits in scanNodes (in auto mode). The call was needed to be able to detect that. It looks like we still don't distriute plans with "soft" limits in the scan node, so I guess it's still needed. CC @solongordon 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Review status: 2 of 14 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/executor.go, line 2092 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbF5YVDaBQ2KZHuveI:-LCVArJCD0l9P9kiLai8:bjgoi5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e632c961b23cbf5f48f364b46d6510ba9f7341/pkg/sql/executor.go#L2092)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The distsql planner did not run plans with limits in scanNodes (in auto mode). The call was needed to be able to detect that. It looks like we still don't distriute plans with "soft" limits in the scan node, so I guess it's still needed. CC @solongordon 
</blockquote></details>

Yes, limit propagation is still needed, but I don't see why it couldn't happen in the caller.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM mod the `shouldUseDistSQL()` interface

---

Review status: 2 of 14 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/conn_executor_exec.go, line 603 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LCZE9ODB0Y4vN-pm56O:-LCZE9ODB0Y4vN-pm56P:b-kktp61) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e298971a7cefd6681703ffc547dfbf75a245378/pkg/sql/conn_executor_exec.go#L603)):*
> ```Go
> 	if !useOptimizer {
> 		useDistSQL, err = shouldUseDistSQL(
> 			ctx, ex.sessionData.DistSQLMode, ex.server.cfg.DistSQLPlanner, planner, nil)
> ```

comment the last arg if it doesn't go away

---

*[pkg/sql/executor.go, line 2092 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbF5YVDaBQ2KZHuveI:-LCZDu3j45DgEcx-hR83:bkfrpq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e632c961b23cbf5f48f364b46d6510ba9f7341/pkg/sql/executor.go#L2092)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Yes, limit propagation is still needed, but I don't see why it couldn't happen in the caller.
</blockquote></details>

I'd say let's move it to the caller then.

---

*[pkg/sql/distsql_wrapper.go, line 60 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LCZD0vz7cbJzSHV_mze:-LCZD0vz7cbJzSHV_mzf:bpn5xft) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bd6dfc91cef743eeed5b20673fce923655e801b/pkg/sql/distsql_wrapper.go#L60)):*
> ```Go
> 			return nil
> 		case <-ctx.Done():
> 			return errors.Wrap(ctx.Err(), "context canceled while writing distSQLWrapper result")
> ```

I think `callbackResultWriter.SetError()` should be called too here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-:-LCZEWSaDpdJWe4aoDdT:b6xs5aw)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 2 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/conn_executor_exec.go, line 603 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LCZE9ODB0Y4vN-pm56O:-LCZWK29-gS1jzUFDF-a:b-glkenb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e298971a7cefd6681703ffc547dfbf75a245378/pkg/sql/conn_executor_exec.go#L603)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

comment the last arg if it doesn't go away
</blockquote></details>

Changed the interface

---

*[pkg/sql/executor.go, line 2092 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LBbF5YVDaBQ2KZHuveI:-LCZWRaF-DQb6h_GdOGk:b81t6oj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e632c961b23cbf5f48f364b46d6510ba9f7341/pkg/sql/executor.go#L2092)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd say let's move it to the caller then.
</blockquote></details>

Done. There was also the subquery check which I moved out and put both behind `prepareForDistSQLSupportCheck`

---

*[pkg/sql/distsql_wrapper.go, line 60 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/24709#-LCZD0vz7cbJzSHV_mze:-LCZWPCQ-sizDgnCmSea:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bd6dfc91cef743eeed5b20673fce923655e801b/pkg/sql/distsql_wrapper.go#L60)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think `callbackResultWriter.SetError()` should be called too here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24709)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
bors r+
